packages/rust/server/src/routes/user.rs

```diff
@@ -1,180 +1,195 @@
+        (
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "unable to create confirmation".to_string(),
+        )
+    })?;
+
+    email_service
+        .send(&format!(
+            "to: {} | body: token: {}",
+            confirmation.email, confirmation.token
+        ))
+        .map_err(|_| {
+            (
+                StatusCode::INTERNAL_SERVER_ERROR,
+                "unable to send email".to_string(),
+            )
+        })?;
 
     Ok(StatusCode::CREATED)
 }
 
 #[derive(Serialize, Deserialize)]
 pub struct Confirm {
-    pub link_id: String,
+    pub token: String,
 }
 
 #[derive(Serialize, Deserialize)]
 pub struct SessionToken {
     pub token: String,
 }
 
-async fn signup_confirm(
+async fn confirm(
     db: Extension<DbPool>,
     Json(payload): Json<Confirm>,
 ) -> Result<(StatusCode, Json<SessionToken>), StatusCode> {
-    let conf = Confirmation::from_link_id(
+    let confirmation = Confirmation::from_token(
         &db.0,
-        &Uuid::parse_str(&payload.link_id).map_err(|_| StatusCode::BAD_REQUEST)?,
+        &Uuid::parse_str(&payload.token).map_err(|_| StatusCode::BAD_REQUEST)?,
     )
     .await
-    .map_err(map_sqlx_err)?;
+    .map_err(|_| StatusCode::NOT_FOUND)?;
 
-    let mut user = User::from_id(&db.0, conf.user_id)
+    confirmation.delete(&db.0).await.map_err(map_sqlx_err)?;
+
+    let mut user = User::from_id(&db.0, confirmation.user_id)
         .await
         .map_err(map_sqlx_err)?;
+    user.email = Some(confirmation.email);
+    user.update(&db.0).await.map_err(map_sqlx_err)?;
+
+    // whenever we change a user's email we should invalidate all tokens
 
-    Confirmation::delete_all(&db.0, user.id)
+    Session::delete_all(&db.0, user.id)
         .await
         .map_err(map_sqlx_err)?;
-    user.email = Some(conf.email);
-    user.update(&db.0).await.map_err(map_sqlx_err)?;
 
-    let mut session = Session {
-        id: 0,
-        user_id: user.id,
-        token: Uuid::new_v4(),
-        created: NaiveDateTime::from_timestamp(0, 0),
-    };
-    session.create(&db.0).await.map_err(map_sqlx_err)?;
+    let token = create_session(&db.0, user.id).await.map_err(map_sqlx_err)?;
 
     Ok((
-        StatusCode::CREATED,
+        StatusCode::OK,
         Json(SessionToken {
-            token: session.token.to_string(),
+            token: token.to_string(),
         }),
     ))
 }
 
 #[derive(Serialize, Deserialize)]
-pub struct SignInJson {
+pub struct Login {
     pub email: String,
     pub password: String,
 }
 
-async fn signin(
+async fn login(
     db: Extension<DbPool>,
-    Json(payload): Json<SignInJson>,
+    Json(payload): Json<Login>,
 ) -> Result<(StatusCode, Json<SessionToken>), StatusCode> {
     let user = User::from_email(&db.0, &payload.email)
         .await
         .map_err(|_| StatusCode::UNAUTHORIZED)?;
 
-    if !user.verify_password(&payload.password) {
+    if !password::verify(&user.password, &payload.password).map_err(|_| StatusCode::UNAUTHORIZED)? {
```

Comment:
This line feels weird, should we just be using something like 
if password::verify(&user.password, &payload.password).[is_err()/?]